### ISSUE TITLE
Grouping features

### ISSUE THREAD

**Issue by andrewesweet** [rmpestano/cukedoctor#145]
**Motivation**
I want to group together features that relate to the same theme or capability and have that grouping reflected in the document hierarchy e.g.

- Theme One
- -  Feature A
- - - Scenario One
- -  Feature B
- - - ...
- Theme Two
- - Feature C
...

**Current Solutions**
Does anyone else have a similar need? Does anyone have some existing solution/workaround to share, please?

The closest I’ve come with “out of the box” cukedoctor is to disable the features section and applying common prefixes to Features in the same Theme, but this is far from idea.

I’ve looked at the Rule keyword in Gherkin 6 (Feature -> Rule -> Scenario), but that does not seem appropriate:
- No cucumber-jvm support yet
- A single Feature must be encapsulated in a single file

**EDIT** - see my subsequent comment for an improved idea than  the one described immediately below.

**How this might be added to cukedoctor**
To introduce Themes, a mechanism would be required to:
- Associate Features with Themes
- Define the order Themes should appear in the document

Two implementation ideas (or a combination thereof) occur to me, both of them clunky:
- Tag abuse
- - Tag each Feature with its Theme e.g. _@theme-MyTheme_
- - - No spaces possible in a tag name, so use underscore as a placeholder and substitute it when rendering.
- - Optionally tag a feature with a theme ordering tag e.g. _@themeorder-42_. The theme associated with that feature will be ordered accordingly.
- - - If conflicting theme ordering arespecified, self-destruct
- - - After partial ordering from tags, apply alphabetic ordering by theme name
- URI abuse
- - Use the relative paths of feature files as a proxy for their desired position in the document hierarchy.
- - Features with a common root path are part of the same theme, with the theme name being the leaf directory’s name
- - Some kind of simple config in document root directory to specify theme order, with secondary alphabetic ordering to ensure a total ordering.

Thoughts and feedback would be appreciated!

**Comment by andrewesweet** [comment]
The next iterations of my thoughts...

**Overview**
A Cukedoctor document can have Sections. We already have some built-in (optional) sections:
- Summary
- Introduction
- Feature (if not used, each individual Feature is rendered as its own Section)

My proposal is to allow _user-defined_ Sections and Sub-Sections. This would, in extremis, allow document authors to create the following document section hierarchy, using all of asciidoc's [six section levels](https://asciidoctor.org/docs/user-manual/#sections):
- Document Title (0)
- - Section (1)
- - - Sub Section (2)
- - - - Feature (3)
- - - - - Rule (4)*
- - - - - -  Scenario/Example (5)
* Placeholder for Gherkin 6's additional of "Rule" keyword. 

**Sections**
To render a Feature in a Section, mark it with the "@section-" tag. For example, this feature:
```
@section-Dinosaurs
Feature: Parasaurolophus
I've got an implausible adornment.
```
would result in the following asciidoc (depending on existing configuration):
```
== Dinosaurs

=== Parasaurolophus
I've got an implausible head adornment.
```

If a Feature is not associated with a Section, it would be rendered as part of the built-in Features Section. If that is disabled, the Feature would be rendered in its own Section. This matches the existing "Section-less" behaviour.

Using only tags, Section names would be very limited (e.g. no spaces). We might also want to write some introduction text in each Section, much as the Introduction Section allows today for the overall document. To allow this, we introduce the idea of a Section Root Feature, using the existing "Stepless Documentation" feature, combined with a "@sectionroot-" tag. For example, if we defined the below feature in addition to the above one:
```
@sectionroot-Dinosaurs
Feature: Dinosaurs: reality or myth?
We all learn to love dinosaurs as children, but did these titans every really exist?
```
would cause the "Dinosaurs" section to now be rendered as follows:
```
== Dinosaurs: reality or myth?
We all learn to love dinosaurs as children, but did these titans every really exist?

=== Parasaurolophus
I've got an implausible head adornment.
```

A Feature with Steps that is tagged as a Section Root should be treated as an error. Whilst one could render it, it is likely a sign that the document author has made a mistake.

Sections can be ordered by applying a "@sectionorder-" tag to a Section Root Feature. Otherwise, Section ordering would be undefined. Applying this tag other than to a Section Root should be treated as an error.

A similar pattern could be followed for Sub-Sections ("@subsection-", "@subsectionroot-", "@subsectionorder-"). If a Feature is associated with a Sub-Section, but not a Section, it will be treated as a Sub-Section of the built-in Features Section. If the built-in Features Section is disabled, this will be treated as an error as it is likely a sign that the document author has made a mistake.

**Comment by rmpestano** [comment]
Hi @andrewesweet,

It looks like an interesting idea but before we go through changing Cukedoctor, have you tried to use [cukedoctor-spi](https://github.com/rmpestano/cukedoctor#692-cukedoctor-spi)?
Maybe we can enhance it to support your idea. 

Note that with the SPI you can even share your customization by just adding your SPI implementation in the classpath.

**Comment by rmpestano** [comment]
We could even release your proposed layout as a separate jar (also project) in this repository and then if user want to use the customized layout he just needs to add your implementation to classpath:

```
 <plugin>
                        <groupId>com.github.cukedoctor</groupId>
                        <artifactId>cukedoctor-maven-plugin</artifactId>
                        <version>${project.parent.version}</version>
                        <configuration>
                            <outputFileName>cukedoctor-documentation</outputFileName>
                            <documentTitle>Cukedoctor Documentation</documentTitle>
                            <featuresDir>target/cukedoctor</featuresDir>
                            <outputDir>cukedoctor</outputDir>
                            <format>all</format>
                            <toc>center</toc>
                            <numbered>true</numbered>
                            <hideSummarySection>true</hideSummarySection>
                            <hideScenarioKeyword>true</hideScenarioKeyword>
                            <sourceHighlighter>coderay</sourceHighlighter>
                            <docVersion>${project.parent.version}</docVersion>
                            <allowUriRead>true</allowUriRead>
                        </configuration>
                        <executions>
                            <execution>
                                <goals>
                                    <goal>execute</goal>
                                </goals>
                                <phase>install</phase>
                            </execution>
                        </executions>
                        <dependencies>
                            <dependency>
                                <groupId>com.github.cukedoctor</groupId>
                                <artifactId>cukedoctor-grouping-layout</artifactId>
                                <version>1.0</version>
                            </dependency>
                        </dependencies>
                    </plugin>
```

**Comment by andrewesweet** [comment]
Hi @rmpestano, using SPI is an excellent call. I should have paid attention to the comprehensive read me!

At first pass, I think the only two changes that would be needed or useful in the core are:
# Create a HeaderRenderer and expose in SPI (this might have been a better change than adding a dedicated attribute to support :stem)
# Enhance CukedoctorDocumentBuilder with some kind of "heading stack", so rather than say docBuilder.sectionTitleLevel1() or .sectionTitleLevel2() etc, you can say .sectionTitleCurrentLevel("myTitle"), .pushSectionTitleLevel(), .popSectionTitleLevel. This could simplify existing could e.g. in CukedoctorFeatureRenderer where we decide to use Level 1 or 2 depending on whether the Features section should be hidden.

I'll have a go and report back.

**Comment by rmpestano** [comment]
Hi, I've just added the HeaderRenderer (#148). 

About the "heading stack" maybe you can add it while implementing the custom layout?

Thanks for the nice idea!

**Comment by andrewesweet** [comment]
Nice, thanks. I'd just started on it.

I hope to have a first pass at the "heading stack" today.

**Comment by andrewesweet** [comment]
Can I please ask a favour, @rmpestano? _CukedoctorConverterTest.shoudlNotRenderFeatureWithSkipDocsTag()_ is failing with me with a NullPointerException in ScenarioTotalizations.addFeatureTotals(), specifically ifeature.getScenarioResults() returns null. This makes sense, since_Feature.scenarioResults_ only gets set by _Feature.processSteps()_, which in turn is only invoked by _Feature.parse(). Given this test does not parse a feature, _Feature.scenarioResults_ will always be null. Therefore, how did this test ever pass?

**Comment by rmpestano** [comment]
Hi @andrewesweet, `Feature.getScenarioResults()` is never called on that test because we're calling **renderFeatures** directly: `Cukedoctor.instance(features, new DocumentAttributes()).
                renderFeatures(features).getDocumentation()`

and `getScenarioResults` is called by `summaryRenderer` only.

**Comment by andrewesweet** [comment]
I missed that in my diff. Thank you, @rmpestano.

**Comment by rmpestano** [comment]
This issue is closed by #149 and will be available in `v3.0`
